translator/source_code/cotrip_translator.py

```diff
@@ -0,0 +1,421 @@
+        end_time = dateutil.parser.parse(end_time_string)
+
+        if end_time < datetime.now(timezone.utc):
+            event_status = "completed"
+    return event_status
+
+
+
+
+
+
+
+# Parse COtrip alert to WZDx
+def parse_alert(alert, callback_function=None):
+    if not validate_alert(alert):
+        if callback_function:    #Note :a call back fucnction , which will trigger every time the invalid data is given
+            callback_function(alert)
+        return None
+    feature = {}
+    geometry = {}
+    geometry['type'] = "LineString"
+    geometry['coordinates'] = parse_polyline(alert['alert:Polyline'])
+
+    feature['type'] = "Feature"
+    properties = {}
+
+    #### I included a skeleton of the message, fill out all required fields and as many optional fields as you can. Below is a link to the spec page for a road event
+    #### https://github.com/usdot-jpo-ode/jpo-wzdx/blob/master/spec-content/objects/RoadEvent.md
+
+    # road_event_id
+    #### Leave this empty, it will be populated by add_ids
+    properties['road_event_id'] = ''
+
+    # Event Type ['work-zone', 'detour']
+    properties['event_type'] = 'work-zone'
+
+    # data_source_id
+    #### Leave this empty, it will be populated by add_ids
+    properties['data_source_id'] = ''
+
+    # start_date
+    properties['start_date'] = reformat_datetime(alert['alert:ReportedTime'])
+
+    # end_date
+    properties['end_date'] = reformat_datetime(alert.get('alert:ExpectedEndTime', ''))
+
+    # start_date_accuracy
+    properties['start_date_accuracy'] = "estimated"
+
+    # end_date_accuracy
+    properties['end_date_accuracy'] = "estimated"
+
+    # beginning_accuracy
+    properties['beginning_accuracy'] = "estimated"
+
+    # ending_accuracy
+    properties['ending_accuracy'] = "estimated"
+
+    # road_name
+    road_name = alert['alert:RoadName']
+    properties['road_name'] = road_name
+
+    # direction
+    Direction_map = {'North': 'northbound', 'South': 'southbound', 'West': 'westbound', 'East': 'eastbound'}
+ 
+    properties['direction'] = Direction_map.get(alert['alert:Direction'])
+
+    # vehicle impact
+
+    properties['vehicle_impact'] = 'unknown' #get_vehicle_impact(alert['alert:ClosureType'])
+
+    # Relationship
+    properties['relationship'] = {}
+
+    # lanes
+    properties['lanes'] = []
+
+    # road_name
+    properties['road_number'] = ""
+
+    # beginning_cross_street
+    properties['beginning_cross_street'] = ""
+
+    # beginning_cross_street
+    properties['ending_cross_street'] = ""
+
+    # event status
+    properties['event_status'] = get_event_status(alert['alert:ReportedTime'], alert.get('alert:ExpectedEndTime'))
+
+    # type_of_work
+    # maintenance, minor-road-defect-repair, roadside-work, overhead-work, below-road-work, barrier-work, surface-work, painting, roadway-relocation, roadway-creation
+    properties['types_of_work'] = []
+
+    # reduced speed limit
+    properties['reduced_speed_limit'] = 25
```

Comment:
This doesn't seem accurate. I see various speed limits in the example data.

Reply:
should be fixed

Reply:
It looks like the `reduced_speed_limit` setting was just removed...does the cotrip not specify if there is a reduced limit? The example data seemed to have speed limit values in it.

Reply:
@abinashkonersman any word on this? It appears to be the same in the `icone_translator.py` where we are had coding the value to 25 for some reason.

Reply:
should be fixed

Reply:
Again, the `reduced_speed_limit` was just removed, but our example data does have speed limit values in it that we should be taking into account.

Reply:
I don't see any reduced_speed_limit in our example data(cotrip_1.json). can you please specify which data i should consider for reduced_speed_limit?

Reply:
It looks like we had reduced speed data in the original [xml file](https://raw.githubusercontent.com/CDOT-CV/Work_Zone/a172aab5ba67ec23ea810bd4ca006e81f55e50ad/translator/sample%20files/cotrip_data/cotrip_data.xml). Do we not have similar data to this in the newer JSON files?

Reply:
you are right and  i found some in  cotrip xml file and working on it

Reply:
created breakout functions for reduced speed limit  which looks for the speed limit in description and publishes it to wzdx message and added required unit test for those function.